### ISSUE TITLE
[BUG] Cannot chain STLBootstrapTransformer, Imputer and BaggingForecaster in one forecast pipeline

### ISSUE THREAD

**Issue by jiayuanteng** [sktime/sktime#5618]
**Describe the bug**
First of all, thanks a lot for putting this package together. It is very good and helpful.

I want to use BaggingForecaster with STLBootstrapTransformer to create boostrapped time series and then use ARIMA as forecaster. 

The issue I ran into is that sometimes boostrapped time series contains NaN. ARIMA does not work with input data that has missing values.

My way of getting around it is to add Impute, i.e. use the nearest number to impute the missing values. 

My question is how should I include impute into the forecasting pipeline? I tried ForecastingPipeline, but I could not feed the imputed and boostrapped time series into the bagging forecaster.

Example code is below. I use NaiveForecaster for simplicity.
**To Reproduce**

```python
<Paste your code here>

from sktime.transformations.series.impute import Imputer

from sktime.forecasting.naive import NaiveForecaster

from sktime.datasets import load_airline

y = load_airline()

bagging =STLBootstrapTransformer(4, sp = 4)  

bagging_forecaster = BaggingForecaster(forecaster=NaiveForecaster(strategy = 'last'))

forecaster  = ForecastingPipeline(

    steps=[

        ("a",bagging),

        ("b", Imputer(method = 'nearest')),

        ("forecaster", bagging_forecaster),

    ]

)

 

forecaster.fit(y=y, fh=[1,2,3])

forecaster.predict()

**Expected behavior**
When you run above code, you can check forecaster.get_fitted_params(), you will see that bagging forecaster used default STLBoostrapTransformer (n_series = 10). 

```

**Additional context**

**Versions**
<details>

</details>

**Comment by benHeid** [comment]
The `ForecastingPipeline` is transforming the exogenous features `X`. I think you would like to transform the target time series itself. Thus, you have to use the `TransformedTargetForecaster` instead of the `ForecastingPipeline`. 

Moreover, I would propose to do the imputation first and then afterwards the bagging.

**Comment by jiayuanteng** [comment]
@benHeid thanks so much for getting back to me. STLboostraptransformer output is multi-index data frame (panel). I tried with TransformedTargetForecaster. It reported error when I call predict.

How can I feed imputed boostrapped time series into BaggingForecaster? Do you mind providing sample code?

Thanks you!

**Comment by benHeid** [comment]
I think your pipeline should look like that:

`pipe = TransformedTargetForecaster([Imputer(), BaggingForecaster(bagging, ARIMA())])`

The `BaggingForecaster` is accepting a `bootstraptransformer` and a `forecaster` as parameters, so I set the bagging you defined previously and an ARIMA model. The first step in the pipeline, is the Imputer. So the data flow would be as follows:

1. Imputer is imputing the time series.
2. BaggingForecaster is using the internal bagging and the ARIMA.

As a shortcut you could also the following notation, which would lead to an equivalent pipeline:

`pipe = Imputer() * BaggingForecaster(bagging, ARIMA())`

Feel free to ask, if you have further questions :) If this solves that issue, please close it :)

**Comment by jiayuanteng** [comment]
@benHeid thanks!

The original time series does not have missing data. Missing data occurs in synthetic time series. Therefore I need to use impute after STLBoostrapTransformer. If I have impute first, it only imputes the original time series. It doesn't help solve the problem.

**Comment by benHeid** [comment]
Oh I see. I am unsure, why there are `nan` values in the bootstrapped time series. Unfortunately, I cannot found any example with `nan` in the bootstrapped time series. So I have to guess how to fix it. But I would suppose that the following line then should work. 

`pipe = TransformedTargetForecaster([Imputer(), BaggingForecaster(bagging, ARIMA())])`

**Comment by jiayuanteng** [comment]
No worries. I will test it out today and report it back.

Perhaps the key question I should have started with is STLBOOSTRAPTRANSFORMER produced NaN values. I have a small time series (24 data points quarterly data). I wonder if that's the reason. I will add data and code to the thread here later.

**Comment by fkiraly** [comment]
@jiayuanteng, interesting, can I understand what you are trying to build? I.e., what the algorithm should do internally, logic-wise?

Do you maybe want to do this?

```python
from sktime.transformations.series.impute import Imputer

from sktime.forecasting.naive import NaiveForecaster

from sktime.datasets import load_airline

y = load_airline()

bootstrap_trafo = STLBootstrapTransformer(4, sp = 4) * Imputer(method="nearest")

bagging_forecaster = BaggingForecaster(
    bootstrap_transformer=bootstrap_trafo,
    forecaster=NaiveForecaster(strategy = 'last'),
)

bagging_forecaster.fit(y=y, fh=[1,2,3])

bagging_forecaster.predict()
```

**Comment by fkiraly** [comment]
also, *should* the `STLBootstrapTransformer` generate `nan`? An example would be interesting, perhaps it is a bug?

**Comment by jiayuanteng** [comment]
@fkiraly thanks for the recommendation. I have a small time series and I want to use BaggingForecaster to improve accuracy. I didn't expect NaN in STLBoostrappedTransformer. I will upload example code later.

Your code didn't work. Does TransformerPipeline handle multiindex? Error message is below. I actually tested TransformerPipeline and make_pipeline before. I could use make_pipeline to chain stlboostraptransformer and impute together to do transformation. But BaggingForecaster doesn't accept make_pipeline as boostrapped_transformer.



_transform output of <class 'sktime.transformations.compose._pipeline.TransformerPipeline'> does not comply with sktime mtype specifications. See datatypes.MTYPE_REGISTER for mtype specifications. Returned error message: ["<class 'pandas.core.indexes.multi.MultiIndex'> is not supported for obj, use one of (<class 'pandas.core.indexes.range.RangeIndex'>, <class 'pandas.core.indexes.period.PeriodIndex'>, <class 'pandas.core.indexes.datetimes.DatetimeIndex'>) or integer index instead.", "obj must be a pandas.Series, found <class 'pandas.core.frame.DataFrame'>", "obj must be a numpy.ndarray, found <class 'pandas.core.frame.DataFrame'>"].

**Comment by fkiraly** [comment]
Hm, interesting, seems like the `TransformerPipeline` does not work for chaining series-to-panel with series-to-series.

Could you also open a new bug report where you build `STLBootstrapTransformer(4, sp = 4) * Imputer(method="nearest")` and pass some time series to it (whatever you passed to provoke the bug), this should in isolation already be a separate bug - at least it would be my guess.

**Comment by jiayuanteng** [comment]
@fkiraly will do.

Regarding STLBoostrappedTransformer that creates NaN. I replicated the code. The reason for NaN is because of inv_cox method. It uses np.exp(np.log1P(lmbda * x) / lmbda) for calculation. When lmbda*x is -1 (or close to negative 1), it means that we want to take natural logarithm of 0 (1+lmbda*x = 0), which causes NaN value.

 

In the example below, the first and the last values are NaN.

 

# before inverse boxcox transformation

x = np.array([ -0.28897117, -0.25959918,-0.25636521, -0.26094224, -0.26413546, -0.26642502])

# lambda is determined by method Guerrero

# X is original time series

# box_cox_transformer_ = BoxCoxTransformer(

#             sp=4, bounds=None, method='guerrero'

#         ).fit(X)

# lmbda = box_cox_transformer_.lambda_

lamdba = 3.75759394981467

# after inverse boxcox transformation

inv_boxcox(x.flatten(),lmbda)

**Comment by fkiraly** [comment]
> I prepared example code but it is not pasting properly on my phone. Let me know if you want to see the code. I can provide it later.

That would be useful. It might not be a bug though, you may like to use `lambda_bounds` in this case, to bound `lambda` away from -1.

**Comment by fkiraly** [comment]
`lambda_bounds` is one of the parameters of `STLBootstrapTransformer`, I think you just set ip.

**Comment by jiayuanteng** [comment]
@fkiraly lambda is not close to -1. Example here lambda is larger than 3. But lambda multiplied by x is negative 1.

**Comment by fkiraly** [comment]
Oh, I see! Your origial series cotains negativ values. In this case the box-cox trafo will be problematic.
Do you perhaps want to use the moving block bootstrap instead?

**Comment by jiayuanteng** [comment]
The original time series have all positive values. After box cox transformation, there are negative values.

**Comment by jiayuanteng** [comment]
@fkiraly I agree that NaN values are not caused by a bug. I will test out TransformerPipeline once that one is fixed. You can close this thread. Thanks a lot for your time😀

**Comment by fkiraly** [comment]
So, in which step exactly do the `nan` appear, and what is exactly the property of the time series, estimator, etc that makes them appear? I assume you've tracked it down, so it would be helpful if you could spell it out for me from this perspective.

**Comment by jiayuanteng** [comment]
@fkiraly  let me give a try. It is probably not a bug - the mathematical calculation makes sense to me. My current solution is to reset random seed or using impute to replace missing values. But I wonder if it is possible to add a condition in _moving_block_boostrap that lambda * x + 1 must be positive.

Original time series is positive and smaller than 1. It is quarterly data. 

```
from statsmodels.tsa.api import STL as _STL
from scipy.special import boxcox
 
from sktime.transformations.series.boxcox import BoxCoxTransformer
 
X = np.array([0.42559563, 0.42497326, 0.45794709, 0.4139578 , 0.40920169,
       0.42456304, 0.4304756 , 0.36854807, 0.38042933, 0.33312965,
       0.42598993, 0.30058192, 0.28701577, 0.29385392, 0.36866835,
       0.27511074, 0.29736489, 0.34672238, 0.39232403, 0.34361043,
       0.22801034, 0.37137277, 0.39635325, 0.35328274, 0.29109913,
       0.33992983])
 
box_cox_transformer_ = BoxCoxTransformer(
            sp=4, bounds=None, method='guerrero'
        ).fit(X)
Xt = boxcox(np.abs(X), box_cox_transformer_.lambda_)
Xt2 = box_cox_transformer_.transform(X)
stl = _STL(endog = Xt,
            period=4).fit()
 
seasonal = pd.Series(stl.seasonal)
resid = pd.Series(stl.resid)
trend = pd.Series(stl.trend)
 
```
I can calculate inverse box cox transformation on the original time series, because lambda * x + 1 is positive
```
Xt2* box_cox_transformer_.lambda_ +1
```
<img width="496" alt="Screenshot 2023-12-15 at 7 09 39 AM" 
src="https://github.com/sktime/sktime/assets/27443463/5ffb5268-f58c-449b-bc01-b8dbb50fb963">

The issue is that when I do a random draw from residual term (transformed). 

```
from sklearn.utils import check_random_state
from scipy.special import inv_boxcox
 
ts = resid
ts_length = len(ts)
ts_index = ts.index
ts_values = ts.values
random_state = 3456
block_length = 8
 
rng = check_random_state(random_state)
mbb_series_col = {}
df_col = {}
df_inv_cox = {}
df_list = []
 
for i in np.arange(0, 6):
  total_num_blocks = int(ts_length / block_length) + 2
  block_origns = rng.choice(
              ts_length - block_length + 1, size=total_num_blocks, replace=False
  )
  mbb_values1 = [
              val for i in block_origns for val in ts_values[i : i + block_length]
              ]
  # remove the first few observations and ensure new series has the
  # same length as the original
  remove_first = rng.choice(block_length - 1)
  mbb_values = mbb_values1[remove_first : remove_first + ts_length]
 
  mbb_series = pd.Series(data=mbb_values, index=ts_index)
 
 
  new_series_bf = mbb_series+seasonal+trend
  new_series_id = f"synthetic_{i}"
  df_col[new_series_id] = new_series_bf
 
  mbb_series_col[new_series_id] = mbb_series
 
  X_shape = X.shape
  new_series = inv_boxcox(new_series_bf.values.flatten(), box_cox_transformer_.lambda_)
  new_series = new_series.reshape(X_shape)
  df_inv_cox[new_series_id]  = new_series
 
  new_df_index = pd.MultiIndex.from_product(
      iterables=[[new_series_id],ts_index],
      names=["series_id", "time_index"],
  )
 
  df_list.append(
              pd.DataFrame(
                  data=new_series, index=new_df_index, columns=['new']
              )
          )
 
 ```

Synthetic_0 has NaN

```
df_inv_cox['synthetic_0']
```

<img width="514" alt="Screenshot 2023-12-15 at 7 14 18 AM" src="https://github.com/sktime/sktime/assets/27443463/8840f4dc-8e44-4601-a479-6eeaf7fce641">

It is because of the random draw of the residual terms. Below shows that the sum of residual, seasonality, and trend, multipled by lambda is smaller than 1. Inverse box cox transformation fails. 

np.round(((mbb_series_col['synthetic_0']+seasonal+trend)*box_cox_transformer_.lambda_+1).values.flatten(), 6)

<img width="542" alt="Screenshot 2023-12-15 at 7 15 36 AM" src="https://github.com/sktime/sktime/assets/27443463/4f918874-b5ec-4b2e-814d-80a3ef24e0be">

If I reset random_state, I will get different draw of residual terms and I *may* get around the NaN values.

**Comment by fkiraly** [comment]
> It is because of the random draw of the residual terms. Below shows that the sum of residual, seasonality, and trend, multipled by lambda is smaller than 1. Inverse box cox transformation fails.

Then it would appear that the draw is not from the correct distribution - should, in theory at least - the draw not end up bigger zero?

Of course, if this is an empirical or pooled estimate, it can happen, so perhaps one may like to bound the resulting draw by a small constant?

I can have a look at the math, if you point me to the equations. The question is, is the implementation faithful and does the original algorithm have this property, or is sth in the implementation not quite right?

**Comment by jiayuanteng** [comment]
I am not deep into moving block boostrap. Perhaps the paper here is going to help? https://www.monash.edu/business/econometrics-and-business-statistics/research/publications/ebs/wp11-14.pdf It is where Prof. Hyndman wrote about bagging forecaster. Section 2.4 has bootstrapping the residuals. 

Draws are from the residual term after box cox transformation on the original time series. The issue is that sometimes the summation of the 3 components together multiplied by lambda can be smaller than -1. 

Below is the code of STLBoostrapTransformer.  (_moving_block_boostrap() + seasonal + trend) is fed into inverse box cox function immediately after the random draws of residual term. There is no ganrantee that  inverse box cox function can work. 

![Screenshot 2023-12-15 at 2 16 30 PM](https://github.com/sktime/sktime/assets/27443463/4b71cf5b-bfe0-4547-89db-bb0e92334c73)

**Comment by jiayuanteng** [comment]
@fkiraly  How about making the following changes to STLBoostrapTransformer as below?

1. I added an if condition that ensures that user passes original time series that can pass inverse box cox function. 
2. I added an if condition after random draw to ensure the summation of residual from random draw, seasonality and trend multipled by lambda is smaller than -1. 

I realize i cannot change the color to highlight the changes. I added ** ** in my edits. 

```
X_transformed = self.box_cox_transformer_.transform(X)

**if X_transformed.self.box_cox_transformer.lambda + 1<=0:
    print('inverse box cox function fails on the original time series')
else:**
 # fit STL on X_transformed series and extract trend, seasonal and residuals
    stl = _STL(
        X_transformed,
        period=self.sp,
        seasonal=self.seasonal,
        trend=self.trend,
        low_pass=self.low_pass,
        seasonal_deg=self.seasonal_deg,
        trend_deg=self.trend_deg,
        low_pass_deg=self.low_pass_deg,
        robust=self.robust,
        seasonal_jump=self.seasonal_jump,
        trend_jump=self.trend_jump,
        low_pass_jump=self.low_pass_jump,
    ).fit(inner_iter=self.inner_iter, outer_iter=self.outer_iter)
    seasonal = pd.Series(stl.seasonal, index=X_index)
    resid = pd.Series(stl.resid, index=X_index)
    trend = pd.Series(stl.trend, index=X_index)

    # time series id prefix
    col_name = _get_series_name(X)

    # initialize the dataframe that will store the bootstrapped series
    if self.return_actual:
        df_list = [
            pd.DataFrame(
                X.values,
                index=pd.MultiIndex.from_product(
                    iterables=[["actual"], X_index],
                    names=["series_id", "time_index"],
                ),
                columns=[col_name],
            )
        ]
    else:
        df_list = []

    # create multiple series
    **for i in range(self.n_series):
    # set the random state
        rng = check_random_state(self.random_state)
        xbb_series = _moving_block_bootstrap(
                    ts=resid,
                    block_length=self.block_length_,
                    replacement=self.sampling_replacement,
                    random_state=rng,
                )
                + seasonal
                + trend
        
        if xbb_series*self.box_cox_transformer.lambda+1<=0:
            rng = check_random_state(self.random_state)
            xbb_series = _moving_block_bootstrap(
                        ts=resid,
                        block_length=self.block_length_,
                        replacement=self.sampling_replacement,
                        random_state=rng,
                    )
                    + seasonal
                    + trend
        
        else:
            
            new_series = self.box_cox_transformer_.inverse_transform(xbb_series)**

        new_series_id = f"synthetic_{i}"
        new_df_index = pd.MultiIndex.from_product(
            iterables=[[new_series_id], new_series.index],
            names=["series_id", "time_index"],
        )

```
